Compile error with xsimd on on ARM64/Apple Silicon

Here is the traceback, unfortunately I cannot share a bigger reproducible due to proprietary code, but feel free to ask for more information.

The xtensor-dependent program compiles fine with xsimd off.

```
error: no viable conversion from returned value of type 'xsimd::batch_bool<float, xsimd::neon64>' to function return type 'simd_return_type<float>' (aka 'xsimd::batch<float>')
```

MCVE:

```cpp
#include <xtensor/xtensor.hpp>

int main() {
    xt::xtensor<double, 2> a = {{1}};
    xt::xtensor<double, 2> b = {{1}};
    auto out = xt::eval(!(xt::isnan(a) | xt::isnan(b)));
}
```

<details><summary>Complete log with xsimd</summary>
```
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/habbasi/Code/xtensor-test/build
Consolidate compiler generated dependencies of target xtensor_test
[ 50%] Building CXX object CMakeFiles/xtensor_test.dir/main.cpp.o
/Users/habbasi/Code/xtensor-test/main.cpp:4:17: error: unknown type name 'float64_t'; did you mean 'float_t'?
    xt::xtensor<float64_t, 2> a = {{1}};
                ^~~~~~~~~
                float_t
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/math.h:44:19: note: 'float_t' declared here
    typedef float float_t;
                  ^
/Users/habbasi/Code/xtensor-test/main.cpp:5:17: error: unknown type name 'float64_t'; did you mean 'float_t'?
    xt::xtensor<float64_t, 2> b = {{1}};
                ^~~~~~~~~
                float_t
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/math.h:44:19: note: 'float_t' declared here
    typedef float float_t;
                  ^
2 errors generated.
make[2]: *** [CMakeFiles/xtensor_test.dir/main.cpp.o] Error 1
make[1]: *** [CMakeFiles/xtensor_test.dir/all] Error 2
make: *** [all] Error 2
❯ rm -rf build/
❯ CMAKE_EXPORT_COMPILE_COMMANDS=1 cmake . -B build/
-- The C compiler identification is Clang 13.0.1
-- The CXX compiler identification is Clang 13.0.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Users/habbasi/mambaforge/envs/cpp/bin/arm64-apple-darwin20.0.0-clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Users/habbasi/mambaforge/envs/cpp/bin/arm64-apple-darwin20.0.0-clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test arch_native_supported
-- Performing Test arch_native_supported - Failed
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/habbasi/Code/xtensor-test/build
❯ cmake --build build/
[ 50%] Building CXX object CMakeFiles/xtensor_test.dir/main.cpp.o
/Users/habbasi/Code/xtensor-test/main.cpp:4:17: error: unknown type name 'float64_t'; did you mean 'float_t'?
    xt::xtensor<float64_t, 2> a = {{1}};
                ^~~~~~~~~
                float_t
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/math.h:44:19: note: 'float_t' declared here
    typedef float float_t;
                  ^
/Users/habbasi/Code/xtensor-test/main.cpp:5:17: error: unknown type name 'float64_t'; did you mean 'float_t'?
    xt::xtensor<float64_t, 2> b = {{1}};
                ^~~~~~~~~
                float_t
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/math.h:44:19: note: 'float_t' declared here
    typedef float float_t;
                  ^
2 errors generated.
make[2]: *** [CMakeFiles/xtensor_test.dir/main.cpp.o] Error 1
make[1]: *** [CMakeFiles/xtensor_test.dir/all] Error 2
make: *** [all] Error 2
❯ cmake --build build/
Consolidate compiler generated dependencies of target xtensor_test
[ 50%] Building CXX object CMakeFiles/xtensor_test.dir/main.cpp.o
[100%] Linking CXX executable xtensor_test
[100%] Built target xtensor_test
❯ rm -rf build/
❯ CMAKE_EXPORT_COMPILE_COMMANDS=1 cmake . -B build/
-- The C compiler identification is Clang 13.0.1
-- The CXX compiler identification is Clang 13.0.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Users/habbasi/mambaforge/envs/cpp/bin/arm64-apple-darwin20.0.0-clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Users/habbasi/mambaforge/envs/cpp/bin/arm64-apple-darwin20.0.0-clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test arch_native_supported
-- Performing Test arch_native_supported - Failed
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/habbasi/Code/xtensor-test/build
❯ cmake --build build/
[ 50%] Building CXX object CMakeFiles/xtensor_test.dir/main.cpp.o
In file included from /Users/habbasi/Code/xtensor-test/main.cpp:1:
In file included from /Users/habbasi/mambaforge/envs/cpp/include/xtensor/xtensor.hpp:20:
In file included from /Users/habbasi/mambaforge/envs/cpp/include/xtensor/xcontainer.hpp:25:
In file included from /Users/habbasi/mambaforge/envs/cpp/include/xtensor/xmath.hpp:28:
In file included from /Users/habbasi/mambaforge/envs/cpp/include/xtensor/xmanipulation.hpp:13:
In file included from /Users/habbasi/mambaforge/envs/cpp/include/xtensor/xbuilder.hpp:30:
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xfunction.hpp:772:16: error: no viable conversion from returned value of type 'xsimd::batch_bool<double, xsimd::neon64>' to function return type 'simd_return_type<double>' (aka 'xsimd::batch<double>')
        return load_simd_impl<align, requested_type, N>(std::make_index_sequence<sizeof...(CT)>(), i);
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xfunction.hpp:830:23: note: in instantiation of function template specialization 'xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>::load_simd<xt::inner_aligned_mode, double, 2UL>' requested here
            .template load_simd<align, requested_type>(i))...);
                      ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xassign.hpp:686:67: note: in instantiation of function template specialization 'xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>::load_simd<xt::inner_aligned_mode, double, 2UL>' requested here
            e1.template store_simd<lhs_align_mode>(i, e2.template load_simd<rhs_align_mode, value_type>(i));
                                                                  ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xassign.hpp:391:47: note: in instantiation of function template specialization 'xt::linear_assigner<true>::run<xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>, xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
                linear_assigner<simd_assign>::run(de1, de2);
                                              ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xassign.hpp:414:20: note: in instantiation of function template specialization 'xt::xexpression_assigner_base<xt::xtensor_expression_tag>::assign_data<xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>, xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
        base_type::assign_data(e1, e2, trivial_broadcast);
                   ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xassign.hpp:205:40: note: in instantiation of function template specialization 'xt::xexpression_assigner<xt::xtensor_expression_tag>::assign_xexpression<xt::xexpression<xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>>, xt::xexpression<xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>>' requested here
            xexpression_assigner<tag>::assign_xexpression(e1, e2);
                                       ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xsemantic.hpp:623:13: note: in instantiation of function template specialization 'xt::assign_xexpression<xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>, xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
        xt::assign_xexpression(*this, e);
            ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xsemantic.hpp:489:37: note: in instantiation of function template specialization 'xt::xcontainer_semantic<xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>>::assign_xexpression<xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
        return this->derived_cast().assign_xexpression(e);
                                    ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xtensor.hpp:555:24: note: in instantiation of function template specialization 'xt::xsemantic_base<xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>>::assign<xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
        semantic_base::assign(e);
                       ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xeval.hpp:48:16: note: in instantiation of function template specialization 'xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>::xtensor_container<xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
        return std::forward<T>(t);
               ^
/Users/habbasi/Code/xtensor-test/main.cpp:6:20: note: in instantiation of function template specialization 'xt::eval<xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
    auto out = xt::eval(!(xt::isnan(a) | xt::isnan(b)));
                   ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:46:11: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'xsimd::batch_bool<double, xsimd::neon64>' to 'const xsimd::batch<double> &' for 1st argument
    class batch : public types::simd_register<T, A>
          ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:46:11: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'xsimd::batch_bool<double, xsimd::neon64>' to 'xsimd::batch<double> &&' for 1st argument
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:58:9: note: candidate constructor not viable: no known conversion from 'xsimd::batch_bool<double, xsimd::neon64>' to 'double' for 1st argument
        batch(T val) noexcept;
        ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:59:9: note: candidate constructor not viable: no known conversion from 'xsimd::batch_bool<double, xsimd::neon64>' to 'std::initializer_list<double>' for 1st argument
        batch(std::initializer_list<T> data) noexcept;
        ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:61:9: note: candidate constructor not viable: no known conversion from 'xsimd::batch_bool<double, xsimd::neon64>' to 'xsimd::batch<double>::register_type' (aka 'float64x2_t') for 1st argument
        batch(register_type reg) noexcept;
        ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:60:18: note: explicit constructor is not a candidate
        explicit batch(batch_bool_type const& b) noexcept;
                 ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/arch/../types/xsimd_neon_register.hpp:101:9: note: candidate function
        XSIMD_DECLARE_SIMD_REGISTER(unsigned long long int, neon, detail::neon_vector_type<unsigned long long int>);
        ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/xsimd_register.hpp:41:9: note: expanded from macro 'XSIMD_DECLARE_SIMD_REGISTER'
        operator register_type() const noexcept { return data; }   \
        ^
In file included from /Users/habbasi/Code/xtensor-test/main.cpp:1:
In file included from /Users/habbasi/mambaforge/envs/cpp/include/xtensor/xtensor.hpp:20:
In file included from /Users/habbasi/mambaforge/envs/cpp/include/xtensor/xcontainer.hpp:25:
In file included from /Users/habbasi/mambaforge/envs/cpp/include/xtensor/xmath.hpp:28:
In file included from /Users/habbasi/mambaforge/envs/cpp/include/xtensor/xmanipulation.hpp:13:
In file included from /Users/habbasi/mambaforge/envs/cpp/include/xtensor/xbuilder.hpp:30:
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xfunction.hpp:1045:16: error: no viable conversion from returned value of type 'xsimd::batch_bool<double, xsimd::neon64>' to function return type 'simd_return_type<double>' (aka 'xsimd::batch<double>')
        return (p_f->m_f.simd_apply)(std::get<I>(m_st). template step_simd<T>()...);
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xfunction.hpp:1052:16: note: in instantiation of function template specialization 'xt::xfunction_stepper<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>::step_simd_impl<double, 0UL, 1UL>' requested here
        return step_simd_impl<T>(std::make_index_sequence<sizeof...(CT)>());
               ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xfunction.hpp:1045:66: note: in instantiation of function template specialization 'xt::xfunction_stepper<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>::step_simd<double>' requested here
        return (p_f->m_f.simd_apply)(std::get<I>(m_st). template step_simd<T>()...);
                                                                 ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xfunction.hpp:1052:16: note: in instantiation of function template specialization 'xt::xfunction_stepper<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>::step_simd_impl<double, 0UL>' requested here
        return step_simd_impl<T>(std::make_index_sequence<sizeof...(CT)>());
               ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xassign.hpp:972:61: note: in instantiation of function template specialization 'xt::xfunction_stepper<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>::step_simd<double>' requested here
                res_stepper.store_simd(fct_stepper.template step_simd<value_type>());
                                                            ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xassign.hpp:400:57: note: in instantiation of function template specialization 'xt::strided_loop_assigner<true>::run<xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>, xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
            strided_loop_assigner<simd_strided_assign>::run(de1, de2);
                                                        ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xassign.hpp:414:20: note: (skipping 2 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
        base_type::assign_data(e1, e2, trivial_broadcast);
                   ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xsemantic.hpp:623:13: note: in instantiation of function template specialization 'xt::assign_xexpression<xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>, xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
        xt::assign_xexpression(*this, e);
            ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xsemantic.hpp:489:37: note: in instantiation of function template specialization 'xt::xcontainer_semantic<xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>>::assign_xexpression<xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
        return this->derived_cast().assign_xexpression(e);
                                    ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xtensor.hpp:555:24: note: in instantiation of function template specialization 'xt::xsemantic_base<xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>>::assign<xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
        semantic_base::assign(e);
                       ^
/Users/habbasi/mambaforge/envs/cpp/include/xtensor/xeval.hpp:48:16: note: in instantiation of function template specialization 'xt::xtensor_container<xt::uvector<bool, xsimd::aligned_allocator<bool, 16>>, 2, xt::layout_type::row_major>::xtensor_container<xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
        return std::forward<T>(t);
               ^
/Users/habbasi/Code/xtensor-test/main.cpp:6:20: note: in instantiation of function template specialization 'xt::eval<xt::xfunction<xt::detail::logical_not, xt::xfunction<xt::detail::bitwise_or, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>, xt::xfunction<xt::math::isnan_fun, const xt::xtensor_container<xt::uvector<double, xsimd::aligned_allocator<double, 16>>, 2, xt::layout_type::row_major> &>>>>' requested here
    auto out = xt::eval(!(xt::isnan(a) | xt::isnan(b)));
                   ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:46:11: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'xsimd::batch_bool<double, xsimd::neon64>' to 'const xsimd::batch<double> &' for 1st argument
    class batch : public types::simd_register<T, A>
          ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:46:11: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'xsimd::batch_bool<double, xsimd::neon64>' to 'xsimd::batch<double> &&' for 1st argument
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:58:9: note: candidate constructor not viable: no known conversion from 'xsimd::batch_bool<double, xsimd::neon64>' to 'double' for 1st argument
        batch(T val) noexcept;
        ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:59:9: note: candidate constructor not viable: no known conversion from 'xsimd::batch_bool<double, xsimd::neon64>' to 'std::initializer_list<double>' for 1st argument
        batch(std::initializer_list<T> data) noexcept;
        ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:61:9: note: candidate constructor not viable: no known conversion from 'xsimd::batch_bool<double, xsimd::neon64>' to 'xsimd::batch<double>::register_type' (aka 'float64x2_t') for 1st argument
        batch(register_type reg) noexcept;
        ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/../types/xsimd_batch.hpp:60:18: note: explicit constructor is not a candidate
        explicit batch(batch_bool_type const& b) noexcept;
                 ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/arch/../types/xsimd_neon_register.hpp:101:9: note: candidate function
        XSIMD_DECLARE_SIMD_REGISTER(unsigned long long int, neon, detail::neon_vector_type<unsigned long long int>);
        ^
/Users/habbasi/mambaforge/envs/cpp/include/xsimd/config/../types/xsimd_register.hpp:41:9: note: expanded from macro 'XSIMD_DECLARE_SIMD_REGISTER'
        operator register_type() const noexcept { return data; }   \
        ^
2 errors generated.
make[2]: *** [CMakeFiles/xtensor_test.dir/main.cpp.o] Error 1
make[1]: *** [CMakeFiles/xtensor_test.dir/all] Error 2
make: *** [all] Error 2
```
</details>

Compiles fine without `xsimd`.

I think I'm seeing the same thing with https://github.com/facebookincubator/velox

```
FAILED: velox/type/CMakeFiles/velox_type.dir/Filter.cpp.o
ccache /usr/bin/c++ -DBOOST_ALL_NO_LIB -DBOOST_ATOMIC_DYN_LINK -DBOOST_CONTEXT_DYN_LINK -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_REGEX_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DFMT_LOCALE -DGFLAGS_IS_A_DLL=0 -I/home/ubuntu/velox/. -I/home/ubuntu/velox/velox/external/xxhash -I/home/ubuntu/velox/third_party/xsimd/include -isystem /home/ubuntu/velox/velox -isystem /home/ubuntu/velox/velox/external -isystem /home/ubuntu/velox/velox/external/duckdb -isystem /home/ubuntu/velox/velox/external/duckdb/tpch/dbgen/include -mcpu=native -std=c++17 -Wno-psabi -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -O3 -DNDEBUG -std=gnu++17 -MD -MT velox/type/CMakeFiles/velox_type.dir/Filter.cpp.o -MF velox/type/CMakeFiles/velox_type.dir/Filter.cpp.o.d -o velox/type/CMakeFiles/velox_type.dir/Filter.cpp.o -c /home/ubuntu/velox/velox/type/Filter.cpp
/home/ubuntu/velox/velox/type/Filter.cpp: In member function ‘virtual xsimd::batch_bool<long int> facebook::velox::common::BigintValuesUsingHashTable::testValues(xsimd::batch<long int>) const’:
/home/ubuntu/velox/velox/type/Filter.cpp:219:57: error: no matching function for call to ‘xsimd::batch<long unsigned int>::batch(xsimd::batch<long int>&)’
  219 |   xsimd::batch<int64_t> indices(xsimd::batch<uint64_t>(x) * M & sizeMask_);
      |                                                         ^
In file included from /home/ubuntu/velox/third_party/xsimd/include/xsimd/xsimd.hpp:61,
                 from /home/ubuntu/velox/./velox/common/base/SimdUtil.h:24,
                 from /home/ubuntu/velox/./velox/type/Filter.h:28,
                 from /home/ubuntu/velox/velox/type/Filter.cpp:25:
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:49:9: note: candidate: ‘template<class ... Ts> xsimd::batch<T, A>::batch(T, T, Ts ...) [with Ts = {Ts ...}; T = long unsigned int; A = xsimd::neon64]’
   49 |         batch(T val0, T val1, Ts... vals) noexcept;
      |         ^~~~~
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:49:9: note:   template argument deduction/substitution failed:
/home/ubuntu/velox/velox/type/Filter.cpp:219:57: note:   candidate expects at least 2 arguments, 1 provided
  219 |   xsimd::batch<int64_t> indices(xsimd::batch<uint64_t>(x) * M & sizeMask_);
      |                                                         ^
In file included from /home/ubuntu/velox/third_party/xsimd/include/xsimd/xsimd.hpp:61,
                 from /home/ubuntu/velox/./velox/common/base/SimdUtil.h:24,
                 from /home/ubuntu/velox/./velox/type/Filter.h:28,
                 from /home/ubuntu/velox/velox/type/Filter.cpp:25:
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:467:12: note: candidate: ‘xsimd::batch<T, A>::batch(xsimd::batch<T, A>::register_type) [with T = long unsigned int; A = xsimd::neon64; xsimd::batch<T, A>::register_type = std::conditional<false, __Int64x2_t, __Uint64x2_t>::type]’
  467 |     inline batch<T, A>::batch(register_type reg) noexcept
      |            ^~~~~~~~~~~
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:467:45: note:   no known conversion for argument 1 from ‘xsimd::batch<long int>’ to ‘xsimd::batch<long unsigned int>::register_type’ {aka ‘std::conditional<false, __Int64x2_t, __Uint64x2_t>::type’}
  467 |     inline batch<T, A>::batch(register_type reg) noexcept
      |                               ~~~~~~~~~~~~~~^~~
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:457:12: note: candidate: ‘xsimd::batch<T, A>::batch(const xsimd::batch_bool<T, A>&) [with T = long unsigned int; A = xsimd::neon64]’
  457 |     inline batch<T, A>::batch(batch_bool<T, A> const& b) noexcept
      |            ^~~~~~~~~~~
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:457:55: note:   no known conversion for argument 1 from ‘xsimd::batch<long int>’ to ‘const xsimd::batch_bool<long unsigned int, xsimd::neon64>&’
  457 |     inline batch<T, A>::batch(batch_bool<T, A> const& b) noexcept
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~^
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:434:12: note: candidate: ‘xsimd::batch<T, A>::batch(T) [with T = long unsigned int; A = xsimd::neon64]’
  434 |     inline batch<T, A>::batch(T val) noexcept
      |            ^~~~~~~~~~~
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:434:33: note:   no known conversion for argument 1 from ‘xsimd::batch<long int>’ to ‘long unsigned int’
  434 |     inline batch<T, A>::batch(T val) noexcept
      |                               ~~^~~
In file included from /home/ubuntu/velox/third_party/xsimd/include/xsimd/xsimd.hpp:61,
                 from /home/ubuntu/velox/./velox/common/base/SimdUtil.h:24,
                 from /home/ubuntu/velox/./velox/type/Filter.h:28,
                 from /home/ubuntu/velox/velox/type/Filter.cpp:25:
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:46:9: note: candidate: ‘xsimd::batch<T, A>::batch() [with T = long unsigned int; A = xsimd::neon64]’
   46 |         batch() = default; ///< Create a batch initialized with undefined values.
      |         ^~~~~
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:46:9: note:   candidate expects 0 arguments, 1 provided
In file included from /home/ubuntu/velox/third_party/xsimd/include/xsimd/xsimd.hpp:61,
                 from /home/ubuntu/velox/./velox/common/base/SimdUtil.h:24,
                 from /home/ubuntu/velox/./velox/type/Filter.h:28,
                 from /home/ubuntu/velox/velox/type/Filter.cpp:25:
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:35:11: note: candidate: ‘constexpr xsimd::batch<long unsigned int>::batch(const xsimd::batch<long unsigned int>&)’
   35 |     class batch : public types::simd_register<T, A>
      |           ^~~~~
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:35:11: note:   no known conversion for argument 1 from ‘xsimd::batch<long int>’ to ‘const xsimd::batch<long unsigned int>&’
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:35:11: note: candidate: ‘constexpr xsimd::batch<long unsigned int>::batch(xsimd::batch<long unsigned int>&&)’
/home/ubuntu/velox/third_party/xsimd/include/xsimd/types/xsimd_batch.hpp:35:11: note:   no known conversion for argument 1 from ‘xsimd::batch<long int>’ to ‘xsimd::batch<long unsigned int>&&’
```

--- a/include/xsimd/arch/generic/xsimd_generic_details.hpp
+++ b/include/xsimd/arch/generic/xsimd_generic_details.hpp
@@ -135,6 +135,51 @@ namespace xsimd
             }
         }
 
+        // some generic fast_cast conversion
+        namespace detail
+        {
+            template <class A>
+            inline batch<uint8_t, A> fast_cast(batch<int8_t, A> const& self, batch<uint8_t, A> const&, requires_arch<generic>) noexcept
+            {
+                return bitwise_cast<uint8_t>(self);
+            }
+            template <class A>
+            inline batch<uint16_t, A> fast_cast(batch<int16_t, A> const& self, batch<uint16_t, A> const&, requires_arch<generic>) noexcept
+            {
+                return bitwise_cast<uint16_t>(self);
+            }
+            template <class A>
+            inline batch<uint32_t, A> fast_cast(batch<int32_t, A> const& self, batch<uint32_t, A> const&, requires_arch<generic>) noexcept
+            {
+                return bitwise_cast<uint32_t>(self);
+            }
+            template <class A>
+            inline batch<uint64_t, A> fast_cast(batch<int64_t, A> const& self, batch<uint64_t, A> const&, requires_arch<generic>) noexcept
+            {
+                return bitwise_cast<uint64_t>(self);
+            }
+            template <class A>
+            inline batch<int8_t, A> fast_cast(batch<uint8_t, A> const& self, batch<int8_t, A> const&, requires_arch<generic>) noexcept
+            {
+                return bitwise_cast<int8_t>(self);
+            }
+            template <class A>
+            inline batch<int16_t, A> fast_cast(batch<uint16_t, A> const& self, batch<int16_t, A> const&, requires_arch<generic>) noexcept
+            {
+                return bitwise_cast<int16_t>(self);
+            }
+            template <class A>
+            inline batch<int32_t, A> fast_cast(batch<uint32_t, A> const& self, batch<int32_t, A> const&, requires_arch<generic>) noexcept
+            {
+                return bitwise_cast<int32_t>(self);
+            }
+            template <class A>
+            inline batch<int64_t, A> fast_cast(batch<uint64_t, A> const& self, batch<int64_t, A> const&, requires_arch<generic>) noexcept
+            {
+                return bitwise_cast<int64_t>(self);
+            }
+        }
+
         namespace detail
         {
             // Generic conversion handling machinery. Each architecture must define

--- a/test/test_conversion.cpp
+++ b/test/test_conversion.cpp
@@ -200,5 +200,49 @@ TEST_CASE_TEMPLATE("[conversion]", B, CONVERSION_TYPES)
         Test.test_u8_casting();
     }
 }
+
+template <class T>
+struct sign_conversion_test
+{
+
+    using unsigned_type = T;
+    using signed_type = typename std::make_signed<T>::type;
+
+    void test_to_signed()
+    {
+        unsigned_type unsigned_value = 3;
+        signed_type signed_value = (signed_type)unsigned_value;
+        xsimd::batch<unsigned_type> unsigned_batch(unsigned_value);
+        auto signed_batch = xsimd::batch_cast<signed_type>(unsigned_batch);
+        CHECK_EQ(unsigned_batch.get(0), unsigned_value);
+        CHECK_EQ(signed_batch.get(0), signed_value);
+    }
+
+    void test_to_unsigned()
+    {
+        signed_type signed_value = 3;
+        unsigned_type unsigned_value = (unsigned_type)signed_value;
+        xsimd::batch<signed_type> signed_batch(signed_value);
+        auto unsigned_batch = xsimd::batch_cast<unsigned_type>(signed_batch);
+        CHECK_EQ(signed_batch.get(0), signed_value);
+        CHECK_EQ(unsigned_batch.get(0), unsigned_value);
+    }
+};
+
+TEST_CASE_TEMPLATE("[conversion]", T, uint8_t, uint16_t, uint32_t, uint64_t)
+{
+    sign_conversion_test<T> Test;
+
+    SUBCASE("to_signed")
+    {
+        Test.test_to_signed();
+    }
+
+    SUBCASE("to_unsigned")
+    {
+        Test.test_to_unsigned();
+    }
+}
+
 #endif
 #endif

